### PR TITLE
add more context to log after a request times out

### PR DESCRIPTION
**Pull request by tkashem**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

when the timeout handler times out a request it writes `504` to the response and then spins up a go routine. The go routine waits for the inner handler to return the result. 
This PR logs the result:
- with request verb and path
- with time elapsed since the timeout

log output:
```
E1221 13:28:44.939372   31341 timeout.go:130] a timeout response has already been sent - time elapsed since: 1.000324432s, GET "/version" result: net/http: abort Handler
```

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
New metric apiserver_request_post_timeout_total that tracks the post-timeout activity of a request after it had been timed out by the apiserver
```

### CODE REVIEW

**Comment by deads2k** [comment]
Given that we are very likely to merge a PR that properly passes and cancels on context and given than primary  usage is to know how much longer a request ran, I think it's more reasonable to make this a metric if we need it to verify correctness of the cancel.  Logging each time is unnecessary and harder to aggregate.

**Comment by smarterclayton** [comment]
As a counter point, we are seeing this error and don't know why - so some level of logging would be useful to identify the cause, although I agree I *also* want a metric.  But metric cardinality will be too limited.

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:266]: @p0lyn0mial had discovered a flake - inner handler (of a request that had been timed out) from a different test can write to the log and this will fail the test. (my change adds a single log entry from such an inner handler)

Maybe there is a different way we can check this?

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:266]: > inner handler (of a request that had been timed out) from a different test can write to the log and this will fail the test

Have you identified that test? This test assumes the other tests are run in serial.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:236]: this comment does not match anymore. Should it?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:279]: comment please what this does

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:285]: So if the call truly hangs forever, with `time.sleep(365*24*time.hour)`, this would stack up a second set of infinite go funcs for logging purposes?

It might be easier to judge correctness here by 
1. creating  struct containing all the various channels so you can comment and give a better than `ch`
2. avoid turning buffered channels into unbuffered channels (it's not obvious why a consumer must exist and should block our producer thread)
2. factoring with a method on that struct to avoid giant inline
3. adding a timeout on waiting for the timeout to finish. Even something order of 30 minutes looks lower risk

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: So I took a stab at refactoring the `finishRequest` function:
- removed the function to its own package
- use a single channel to propagate the result from the child goroutine to the parent.

This refactor allows me to inject fakes so I can test the post-timeout behavior. I wrote unit tests to verify that the post-timeout `receiver` goroutine we spin up does not block, it waits for the child goroutine to return a value. So now the post-timeout logic is properly tested and verified.

@deads2k @sttts  this refactor seems out of scope of this PR, let me know if you want me to open a new PR with just the refactor.


> avoid turning buffered channels into unbuffered channels (it's not obvious why a consumer must exist and should block our producer thread)

Now, we spin up a post-timeout receiver goroutine that is waiting to receive, so there is no need to make the channel buffered. Also, the test I wrote ensures that the child goroutine does not block on sending after the request times out.

**Comment by tkashem** [comment]
I have opened a separate PR with just the refactor for `FinishRequest` - https://github.com/kubernetes/kubernetes/pull/100523. 

/hold

**Comment by tkashem** [comment]
(rebased after #100523 has merged) 

/hold cancel

**Comment by tkashem** [comment]
> adding a timeout on waiting for the timeout to finish. Even something order of 30 minutes looks lower risk

done, @deads2k please take a look when you have time

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:124]: this isn't actually started, this is the time the request timedout, right?

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:124]: yes, after the request had timed out, this the `startedAt` for the post-timeout tracker. `timedOutAt` reads better?

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go]: @deads2k I renamed to `timedOutAt`.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:148]: why doesn't this have a metric?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:152]: do we want to know how long we actually are?

**Comment by deads2k** [comment]
@tkashem work with @logicalhan (I he's the metrics guy) about the metric shape

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go:124]: why does this block forever.  I would expect to see similar: "this gofunc must end" logic.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:77]: `30 minutes` - wow that's a long timeout, how did we end up with that number?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:132]: `resultCh` is buffered `resultCh := make(chan *result, 1)`, so the sender will not block sending the `result`, we just are not going to do any post-timeout logging.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:159]: what if `r.reason` is nil? will it panic?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:77]: instead of waiting infinitely, we are waiting for certain threshold. This is so that if there is such a goroutine that never returns we would know about it.  I am adding this to clarify:
```
const (
	// how much time the post-time logger goroutine will wait for the sender
	// (child goroutine executing ResultFunc) to send a result after the request.
	// had timed out.
	postTimeoutLoggerWait = 30*time.Minute
)
```

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:159]: it should not because of `%v`

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:159]: actually, I have a test for this and here is the output:
```
E0413 09:51:23.299443  127665 finisher.go:167] FinishRequest: post-timeout activity - time-elapsed: 1.702µs, panicd: false, panic-reason: <nil>
```

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:159]: thanks, now I recall that it would dereference the value stored at the pointer but it won't use it

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher_test.go
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher_test.go:237]: isn't `finishRequest` blocking?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher_test.go
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher_test.go:199]: ah, okay this is how we terminate `finishRequest` function

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:163]: okay, I am recording a metric for this now.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go]: It sounds useful, I added logging for it.

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go:125]: this is in keeping with the current implementation of the timeout filter. If I make this change I would like to add a test for it, and a proper test necessitates a refactor of the timeout filter. 
@deads2k I would like to do a follow up PR for this. Let me know if you want me to address it in this PR.

**Comment by tkashem** [comment]
@logicalhan please review the metric I am proposing here and let me know if you have any suggestions.

/assign @logicalhan

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:80]: Why is this 30 minutes?

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go:128]: If we're going to be storing bools here, we shouldn't be storing the string 'unknown' in the same field..

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:80]: after the request had timed out, we wait up to 30 minutes for the sender to return a result instead of waiting indefinitely. `30m` seems to be a good starting point.

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go:128]: On second thought, I collapsed both into one label `reason`, possible values: `panicd`, `error`, `ok` and `unknown`

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go]: isn't a panic just a type of error?

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go]: @logicalhan  I think it's useful to preserve the distinction (or context) here - the metric can answer the question whether the child goroutine executing `ResultFunc` panicked or returned an `error`.  let me know if you have any concern with the cardinality here. 

Also, this counter is not of O(number of requests in flight), but is of O(request timed out).

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go]: okay, then i would use 'panic' instead of 'panicd'

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go]: @logicalhan done, PTAL

**Comment by tkashem** [comment]
@logicalhan please take a look when you have a moment.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:79]: It seems like a huge period of time.. What kind of request takes 30 minutes to process?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:167]: it's probably cleaner to declare these as consts.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:223]: `pending` seems more appropriate.

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:357]: this comment is also no longer correct.

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:261]: this comment is no longer correct, since you seem to be okay with 2 lines.

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go]: I fixed the flake, instead of relying on the number of lines in the log, the test now looks for specific patterns to figure out if the stack trace is logged.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: yes, agreed.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:79]: @logicalhan  The apiserver will timeout a request after `60s`, the caller is going to receive a `504`.  This post-timeout receiver kicks in after the request had been timed out, it waits for `30m` to receive a result from the executing request handler.

> What kind of request takes 30 minutes to process?
No request should take 30m to process. The purpose of the post-timeout receiver is to track how long a request handler is active after the request had been timed out. But we don't want to wait forever, `30m` seems to be good starting point.

One example, due to underlying network conditions (stale socket) an http response read/write could block for up to 15 to 20 minutes (depending on kernel settings).

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:264]: have you identified a test that fails on the previous logic?
from what I have seen there no tests in this pkg that would use `FinishRequest` method.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:79]: `30m` seems high since the default timeout is `60s` anything beyond that is concerning. I would expect the context to propagate the cancellation signal and to clean up resources.

From what I have seen `FinishRequest` is used to wrap the storage calls. I would not expect these calls to last 30 minutes.

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:264]: @p0lyn0mial it's coming from this test - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go#L136

**Comment by tkashem** [comment]
@p0lyn0mial @logicalhan I reduced the post-timeout receiver wait to `5m`.

**Comment by tkashem** [comment]
@logicalhan please let me know if you have any further feedback on the metrics shape.

**Comment by tkashem** [comment]
@logicalhan friendly ping

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:231]: I prefer status as opposed to reason to stay consistent with the codebase

**Review by logicalhan — COMMENTED** [review]
other than the label name, the metrics changes lgtm

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go]: @logicalhan I changed the label to `status` as suggested.

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

from the metrics side of things

**Pull request merged**
